### PR TITLE
Pixeldoubling for tilesets

### PR DESCRIPTION
**Pull request by SimonPurdie**
Adds a new option for in tile_config.json to allow tilesets to specify a pixeldoubling multiplier. The tileset will be resized by that multiplier when it is drawn.

This allows the removal of image files for Retrodays20 (which are pixeldoubled copies of Retrodays10 ).

It also means the files for hitbutton_iso can be reduced to native resolution (reducing the spritesheet sizes by 75%).

### CODE REVIEW

File: src/cata_tiles.cpp [hunk 1]
```diff
@@ -416,8 +416,8 @@ int cata_tiles::load_tileset(std::string img_path, int R, int G, int B, int spri
 }
 
 void cata_tiles::set_draw_scale(int scale) {
-    tile_width = default_tile_width * scale / 16;
-    tile_height = default_tile_height * scale / 16;
+    tile_width = default_tile_width * tile_pixelscale * scale / 16;
+    tile_height = default_tile_height * tile_pixelscale * scale / 16;
```
Comment (sparr) [inline src/cata_tiles.cpp]: We're relying on SDL to do the scaling for us during blitting, as is already done with zooming?

Comment (hitbutton) [inline src/cata_tiles.cpp]: Yes. That is the only place in the code that "tile_pixelscale" is checked, which is also the function called to set zoom level. The two are functionally equivalent.

**Comment by sparr** [comment]
Other than causing a problem for people who previously had it selected if we delete it, does retrodays20 need to exist at all any more?

**Comment by hitbutton** [comment]
Yes. It still has exactly the same function it had before.

**Comment by sparr** [comment]
doing the doubling at render time, when we blit, same as zooming, is elegant.

however, it precludes having different spritesheets in the same tileset render at different sizes. would that be a useful feature here?

**Comment by sparr** [comment]
(also, storing it tileset-wide is a factor in the above)

**Comment by hitbutton** [comment]
> it precludes having different spritesheets in the same tileset render at different sizes. would that be a useful feature here?

It could conceivably be useful. But none of the existing tilesets require a feature like that and I'm not sure how likely such a feature is to be noticed / used by any spritesheet authors without any existing examples of its use.

I'd advocate waiting for a use case first.

**Comment by chaosvolt** [comment]
Hmm. Very peculiar. Would this mean I no longer need to retain both versions of MShock Modded?

**Comment by DanmakuDan** [comment]
It looks like the value is only an integer, so it doesn't scale 24 to 32 (1.5).

**Comment by chaosvolt** [comment]
Ah, right. Oh well. Another clever feature that I'll fail to use properly. o3o

**Comment by hitbutton** [comment]
It's an interesting suggestion. There's no reason that pixelscale shouldn't be able to accept a float (ultimately tile_width and tile_height would be truncated to ints after conversion).

I just tried out using MShock32Modded with a pixelscale of 0.75, and it looks slightly different to MShock24Modded. When you do your manual conversion do you work on the 24 set and enlarge, or work on the 32 set then shrink?

**Comment by chaosvolt** [comment]
I've done a mix of both methods, and in some cases I remade the sprite anew if I was able to remember the editing steps.

Even then, both resize methods tend to cause some smudging and color blending that leads to different appearances. Most common with anything that has a border, whereas terrain sprites that occupy the entire space on the grid tend to resize with the least trouble.

**Comment by hitbutton** [comment]
Obviously an automatic rescaling would have the same issues with smudging and color blending, so an automatic conversion probably wouldn't be something you'd want to do.

For reference here's what I got by scaling 32 down to 24 ingame:
image

**Comment by hitbutton** [comment]
> It looks like the value is only an integer, so it doesn't scale 24 to 32 (1.5).

Just thought I'd point out thats actually 1.3r (or 4/3) :)

**Comment by chaosvolt** [comment]
Hmm. I am seeing some issues with the terrain and player sprites, which weren't my doing. ;w;

**Comment by hitbutton** [comment]
Yes, if you want pixel perfect precision, you can't enlarge an image by anything other than a power of two.

**Comment by sparr** [comment]
If we allowed per-spritesheet (rather than per-tileset) scaling, you could scale 32 down to 24 (0.75) for the defaults in the 24px tileset, and then provide a sheet of hand-made 24px sprites where it matters.

**Comment by sparr** [comment]
However, this is all conceptual. The PR as-is seems worth merging. I'll test it when I can, if no one else gets to it first.

**Comment by hitbutton** [comment]
Doh, you're quite right.

**Comment by sparr** [comment]
> Powers of 2 are more efficient though because you can avoid multiplication and division by using left and right shifts instead.

The scaling is being done in hardware while blitting the textures. There is no arithmetic being done in code for this.

**Comment by chaosvolt** [comment]
Hmm. Will still be interesting to see.

**Comment by kevingranade** [comment]
Why not resize at load instead of at render time? Tiles are already running quite slowly on many systems, so it seems pretty egregious to do rescaling on every draw operation instead of once on load.

**Comment by hitbutton** [comment]
I'd have to do some testing with software rendering, but I don't think it should be expensive at all. If I understand the SDL render process correctly then the "rescaling" isn't a complex process, and the texture itself is not changing in memory. I'll do some tests, but I suspect that making the tiles 4x bigger in memory than they need to be wouldn't actually boost performance at all.

This is already what's done when using any zoom level ingame other than the default. So the same issue would arise at any level of zoom other than the default, unless we resize tilesets in memory when we change zoom too.

**Comment by kevingranade** [comment]
It's the same as the zoom level, which is very noticeably slow on my
system, this isn't theory, it *is* slow.

**Comment by sparr** [comment]
Is zooming *in* slow? Zooming out will, of course, be slower, since it means more tiles are drawn.

**Comment by hitbutton** [comment]
Does Retrodays20 run faster for you than Retrodays10 zoomed in one level?

Using software rendering I can't discern any performance difference between the two.

Although god tiles is slow in general without hardware acceleration, you aren't kidding.

**Comment by kevingranade** [comment]
Actually I just got back to my computer and re-tested with zooming, and it doesn't have the slowdown anymore, it's even ok zoomed all the way out so I can see the entire visible area.  So that's awesome, I might actually use zooming sometimes now.

**Pull request merged**